### PR TITLE
Refactor CxxPrepareForLinkStep for a reuseable & more generic step

### PR DESCRIPTION
**Pull request by nguyentruongtho**
Implement `CxxPrepareForLinkStep` as a composite step of 2 `CxxWriteArgsToFileStep` steps, as requested to simplify https://github.com/facebook/buck/pull/917.

### CODE REVIEW

File: src/com/facebook/buck/cxx/CxxPrepareForLinkStep.java [hunk 1]
```diff
@@ -20,126 +20,65 @@
...
-      }
-    }
-  }
+    ImmutableList<Arg> allArgs = ImmutableList.<Arg>builder()
+        .addAll(StringArg.from(linker.outputArgs(output.toString())))
+        .addAll(args)
+        .addAll(linkerArgsToSupportFileList)
+        .build();
 
-  private void cleanupPreviousFiles() throws IOException {
```
Comment (Coneko) [inline src/com/facebook/buck/cxx/CxxPrepareForLinkStep.java]: You need to maintain this functionality though. Move it into `CxxLink` by adding a new `RmStep` there.

**Comment by Coneko** [comment]
Very nice!

File: src/com/facebook/buck/cxx/CxxPrepareForLinkStep.java (hunk 1 above)
Comment (nguyentruongtho) [inline src/com/facebook/buck/cxx/CxxPrepareForLinkStep.java]: image

**Comment by Coneko** [comment]
Tests are failing.

**Comment by nguyentruongtho** [comment]
`appveyor` uses windows for testing, nice!

**Comment by Coneko** [comment]
Tests are still failing.

Can you please run the tests locally before submitting the PR?

**Comment by nguyentruongtho** [comment]
I think it was passing, on Appveyor at least. Travis hung for some reasons.

I normally run related tests locally but not all the tests we have in buck. The last failure test was because it was run on windows (couldn't help )

I rather let CI run all tests in because everytime I do it locally, it eats up 100% cpu and ram on my machine, takes 15-20 mins at least #movefast :)

**Comment by nguyentruongtho** [comment]
I can't retrigger Travis, can you do it?

**Comment by Coneko** [comment]
Restarting the job doesn't seem to be helping.

**Comment by nguyentruongtho** [comment]
There is a problem with travis for mac builds
https://www.traviscistatus.com/

I've tested it locally, everything is good. Possible to merge in? You guys did it quite a few time already, didn't you?

**Comment by Coneko** [comment]
`com.facebook.buck.crosscell.InterCellIntegrationTest.xCellCxxLibraryBuildsShouldBeHermetic` is failing.

Did it not fail for you locally?

**Comment by nguyentruongtho** [comment]
not, it wasn't! I had two fails from `com.facebook.buck.apple.*` due to my local setting.

**Comment by Coneko** [comment]
Still getting `FAILURE com.facebook.buck.crosscell.InterCellIntegrationTest xCellCxxLibraryBuildsShouldBeHermetic: Expected exit code 0 but was 1.`:

```
[2016-10-14 06:58:44.875][warn ][tid:984][com.facebook.buck.event.listener.LoggingBuildListener] g++: no input files

[2016-10-14 06:58:44.876][warn ][tid:984][com.facebook.buck.event.listener.LoggingBuildListener] BUILD FAILED: //:bin#binary failed with exit code 1:
c++ link
stderr: g++: no input files
```

**Comment by nguyentruongtho** [comment]
Let me check, sounds like os dependent problem again.

**Comment by nguyentruongtho** [comment]
Is this should be tested on Linux machine? Feel like some failure of TravisCI last week affected the result of the PR. I'm rebasing and pushing it just for checking.

**Comment by Coneko** [comment]
This wasn't on Travis, it was on our internal CI. I can't repro it either because I don't have gcc.

**Comment by nguyentruongtho** [comment]
I have gcc on my mac, but it passed, let me try get a linux box and try.

**Comment by Coneko** [comment]
Are you sure you were using actual gcc and not the apple gcc that just calls clang?

**Comment by nguyentruongtho** [comment]
sorry for the delay, working to fix the framework resource propagation, almost finish. I'll get back to this one after that gets done.

**Comment by nguyentruongtho** [comment]
I've cast a spell on Travis 👍 
image

**Comment by nguyentruongtho** [comment]
Travis resists to approve my PR, even it is not my fault this time. Well played, Travis.

```
/home/travis/build/facebook/buck/src/com/facebook/buck/lua/AbstractNativeExecutableStarter.java:172: error: no suitable method found for filter(java.lang.Iterable<capture#1 of ? extends com.facebook.buck.cxx.CxxPreprocessorDep>,com.google.common.base.Predicate<java.lang.Object>)
        Iterables.filter(deps, Predicates.not(BuildRule.class::isInstance))) {
                 ^
    method com.google.common.collect.Iterables.<T>filter(java.lang.Iterable<T>,com.google.common.base.Predicate<? super T>) is not applicable
      (cannot infer type-variable(s) T,T
        (argument mismatch; com.google.common.base.Predicate<capture#1 of ? extends com.facebook.buck.cxx.CxxPreprocessorDep> cannot be converted to com.google.common.base.Predicate<? super capture#1 of ? extends com.facebook.buck.cxx.CxxPreprocessorDep>))
    method com.google.common.collect.Iterables.<T>filter(java.lang.Iterable<?>,java.lang.Class<T>) is not applicable
      (no instance(s) of type variable(s) T exist so that com.google.common.base.Predicate<T> conforms to java.lang.Class<T>)
Errors: 1. Warnings: 0.
```

**Comment by marcinkosiba** [comment]
> Travis resists to approve my PR, even it is not my fault this time. Well played, Travis.

I just fixed it. It was a jdk version problem. Try rebasing and it should be fine (well, or at least fail on something else)

**Comment by nguyentruongtho** [comment]
> I just fixed it. It was a jdk version problem. Try rebasing and it should be fine (**well, or at least fail on something else**)

image

**Pull request closed without merging**